pkg/envoy/agent.go

```diff
@@ -105,13 +138,71 @@ func (a *Agent) terminate() {
 	if e != nil {
 		log.Warnf("Error in invoking drain listeners endpoint %v", e)
 	}
-	log.Infof("Graceful termination period is %v, starting...", a.terminationDrainDuration)
-	time.Sleep(a.terminationDrainDuration)
-	log.Infof("Graceful termination period complete, terminating remaining proxies.")
-	a.abortCh <- errAbort
+	log.Infof("Agent draining proxy for %v, then waiting for active connections to terminate...", a.minDrainDuration)
+	time.Sleep(a.minDrainDuration)
+	log.Infof("Termination drain duration period is %v, checking for active connections...", a.terminationDrainDuration)
+	go a.waitForDrain()
+	select {
+	case <-a.drainCh:
+		log.Info("There are no more active connections. terminating proxy...")
+		a.abortCh <- errAbort
+	// TODO: remove terminationDrainDuration and rely on "terminationGracefulPeriodSeconds" of pod?
+	case <-time.After(a.terminationDrainDuration):
+		log.Info("Termination period complete, terminating proxy...")
+		a.abortCh <- errAbort
+		close(a.drainCh)
+	}
 	log.Warnf("Aborted all epochs")
 }
 
+func (a *Agent) waitForDrain() {
+	for {
+		select {
+		case <-time.After(activeConnectionCheckDelay):
+			if a.activeProxyConnections() == 0 {
+				close(a.drainCh)
+				return
+			}
+		case <-a.drainCh:
+			return
+		}
+	}
+}
+
+func (a *Agent) activeProxyConnections() int {
+	activeConnectionsURL := fmt.Sprintf("http://%s:%d/stats?usedonly&filter=downstream_cx_active", a.localhost, a.adminPort)
```

Comment:
very very nit: IIRC regex performance here has a big impact. It is only on shutdown so not a huge deal, but does adding things like `downstream_cx_active$` make it cheaper?

Reply:
I recall something in the past replacing a `.` with a `\.` or similar made it like 4x faster or something crazy is all.

Reply:
Will take a look